toolchain/parse/typed_nodes.h

```diff
@@ -0,0 +1,781 @@
+
+// A tuple literal: `()`, `(a, b, c)`, or `(a,)`.
+struct TupleLiteral {
+  static constexpr auto Kind =
+      NodeKind::TupleLiteral.Define(NodeCategory::Expr);
+  ExprOpenParenId left_paren;
+  CommaSeparatedList<AnyExpr, TupleLiteralCommaId> elements;
+};
+
+// The opening portion of a call expression: `F(`.
+//
+// TODO: Consider flattening this into `CallExpr`.
+struct CallExprStart {
+  static constexpr auto Kind = NodeKind::CallExprStart.Define();
+  AnyExpr callee;
+};
+
+using CallExprComma = LeafNode<NodeKind::CallExprComma>;
+
+// A call expression: `F(a, b, c)`.
+struct CallExpr {
+  static constexpr auto Kind = NodeKind::CallExpr.Define(NodeCategory::Expr);
+  CallExprStartId start;
+  CommaSeparatedList<AnyExpr, CallExprCommaId> arguments;
+};
+
+// A simple member access expression: `a.b`.
+struct MemberAccessExpr {
+  static constexpr auto Kind =
+      NodeKind::MemberAccessExpr.Define(NodeCategory::Expr);
+  AnyExpr lhs;
+  // TODO: Figure out which nodes can appear here
+  NodeId rhs;
+};
+
+// A simple indirect member access expression: `a->b`.
+struct PointerMemberAccessExpr {
+  static constexpr auto Kind =
+      NodeKind::PointerMemberAccessExpr.Define(NodeCategory::Expr);
+  AnyExpr lhs;
+  // TODO: Figure out which nodes can appear here
+  NodeId rhs;
+};
+
+// A prefix operator expression.
+template <const NodeKind& KindT>
+struct PrefixOperator {
+  static constexpr auto Kind = KindT.Define(NodeCategory::Expr);
+  AnyExpr operand;
+};
+
+// An infix operator expression.
+template <const NodeKind& KindT>
+struct InfixOperator {
+  static constexpr auto Kind = KindT.Define(NodeCategory::Expr);
+  AnyExpr lhs;
+  AnyExpr rhs;
+};
+
+// A postfix operator expression.
+template <const NodeKind& KindT>
+struct PostfixOperator {
+  static constexpr auto Kind = KindT.Define(NodeCategory::Expr);
+  AnyExpr operand;
+};
+
+// Literals, operators, and modifiers
+
+#define CARBON_PARSE_NODE_KIND(...)
+#define CARBON_PARSE_NODE_KIND_TOKEN_LITERAL(Name, ...) \
+  using Name = LeafNode<NodeKind::Name, NodeCategory::Expr>;
+#define CARBON_PARSE_NODE_KIND_TOKEN_MODIFIER(Name, ...) \
+  using Name##Modifier =                                 \
+      LeafNode<NodeKind::Name##Modifier, NodeCategory::Modifier>;
+#define CARBON_PARSE_NODE_KIND_PREFIX_OPERATOR(Name, ...) \
+  using PrefixOperator##Name = PrefixOperator<NodeKind::PrefixOperator##Name>;
+#define CARBON_PARSE_NODE_KIND_INFIX_OPERATOR(Name, ...) \
+  using InfixOperator##Name = InfixOperator<NodeKind::InfixOperator##Name>;
+#define CARBON_PARSE_NODE_KIND_POSTFIX_OPERATOR(Name, ...) \
+  using PostfixOperator##Name =                            \
+      PostfixOperator<NodeKind::PostfixOperator##Name>;
+#include "toolchain/parse/node_kind.def"
+
+// The first operand of a short-circuiting infix operator: `a and` or `a or`.
+// The complete operator expression will be an InfixOperator with this as the
+// `lhs`.
+// FIXME: should this be a template?
```

Comment:
Do you want this answered before the PR lands? If not:
```suggestion
// TODO: Should this be a template?
```
I don't really have a strong opinion here. It's not much duplication compared to the overhead of writing a template and two aliases. I'd be happy to say that we don't make this into a template until / unless we want to write generic code to cover both cases at once (eg, in check).

Reply:
Updated comment.